wallet: Construct ScriptPubKeyMans with all data rather than loaded progressively

achow101 opened the pull request:
Instead of constructing ScriptPubKeyMans with no data, and then loading data as we find it, we should gather everything first and then load it all on construction. If there actually is no data and we want to setup generation, then that should also occur in a constructor rather than afterwards.

This change is only applied to DescriptorScriptPubKeyMan and ExternalSignerScriptPubKeyMan, and should be done for any ScriptPubKeyMans added in the future. I don't think it's really worth it to do this for LegacyScriptPubKeyMan since it would make loading performance worse (or cause layer violations) and it's (supposed to be) going away soon.

S3RK commented on src/wallet/wallet.cpp:
```diff
@@ -391,40 +391,18 @@ std::shared_ptr<CWallet> CreateWallet(WalletContext& context, const std::string&
     }
 
+    // Unset the blank flag if not specified by the user
+    if (!create_blank) {
```
Is this change of behaviour or not?

achow101 replied:
It is not.

achow101 committed:
wallet: Consolidate generation setup callers into one function

achow101 committed:
wallet: Load everything into DescSPKM on construction

Instead of creating a DescSPKM that is then progressively loaded, we
should instead create it all at once in a factory function when loading.

achow101 committed:
wallet: include keys when constructing DescriptorSPKM during import

When importing a descriptor, all of the descriptor data should be
provided at the same time in the constructor.

achow101 committed:
wallet: Construct ExternalSignerSPKM with the new descriptor

Instead of constructing then setting the descriptor with
SetupDescriptor, just pass in that descriptor to the constructor.

achow101 committed:
wallet: Setup new autogenerated descriptors on construction

Instead of having a caller use SetupDescriptorGeneration, just have a
constructor that takes those arguments and sets up the descriptor with
the autogenerated key.

achow101 committed:
fuzz: Skip adding descriptor to wallet if it cannot be expanded

achow101 commented:
Added a commit so that the fuzzer will skip the type of inputs that was causing the previous failure. This should be safe as that failure is not reachable in normal usage. It can only be reached if the user's wallet is corrupted, and in that case, the runtime error would propagate up to them so there would not be a crash.

rkrux reviewed (commented):
I seem to be getting onboard with this idea as I can notice readability improvements on a cursory glance. Do you anticipate any performance improvements as well?

The last sentence related to legacy SPKMs of the last paragraph in the PR description can be removed now.

Will do full review soon.

rkrux reviewed (commented):
Concept ACK 93c3729ad5adc02f6d87ec8daa1a1952d7c41009

This PR makes `DescriptorScriptPubKeyMan` class RAII compliant by adding 4 constructors that are responsible for allocating all the resources (eg: memory) required by this class during initialisation itself. I feel that it improves the readability quite a lot as the reader can easily gauge what all is needed by this class at one place and there is no need for multiple scattered class-write function calls like the one below that used to appear before.

```diff
- WITH_LOCK(desc_spk_man->cs_desc_man, desc_spk_man->AddDescriptorKeyWithDB(batch, key, key.GetPubKey()));
- desc_spk_man->TopUpWithDB(batch);
```

I noticed that this style now allows us to add more error checking that makes the code more robust imo.
```cpp
Error: Wallet contains both unencrypted and encrypted keys
```

Perhaps a Refactoring label is required in the PR?

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -420,30 +420,18 @@ std::shared_ptr<CWallet> CreateWallet(WalletContext& context, const std::string&
+    }
+
     // Encrypt the wallet
     if (!passphrase.empty() && !(wallet_creation_flags & WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
```
Done

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3588,6 +3577,18 @@ void CWallet::SetupDescriptorScriptPubKeyMans()
+    }
+    SetupDescriptorScriptPubKeyMans();
+    return true;
+}
```
Made `void`

achow101 commented on src/wallet/walletdb.cpp:
```diff
@@ -903,16 +895,17 @@ static DBErrors LoadDescriptorWalletRecords(CWallet* pwallet, DatabaseBatch& bat
         num_keys = key_res.m_records;
 
         // Get encrypted keys
+        std::map<CKeyID, std::pair<CPubKey, std::vector<unsigned char>>> ckeys;
```
Done

achow101 commented on src/wallet/walletdb.cpp:
```diff
@@ -860,15 +853,14 @@ static DBErrors LoadDescriptorWalletRecords(CWallet* pwallet, DatabaseBatch& bat
+        desc.cache = cache;
 
         // Get unencrypted keys
+        std::map<CKeyID, CKey> keys;
```
Done

achow101 commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -173,15 +173,18 @@ static const std::unordered_set<OutputType> LEGACY_OUTPUT_TYPES {
+using KeyMap = std::map<CKeyID, CKey>;
+using CryptedKeyMap = std::map<CKeyID, std::pair<CPubKey, std::vector<unsigned char>>>;
+using ScriptPubKeyMap = std::map<CScript, int32_t>; // Map of scripts to descriptor range index
+using PubKeyMap = std::map<CPubKey, int32_t>; // Map of pubkeys involved in scripts to descriptor range index
```
I think leaving them here is fine.

achow101 commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -173,15 +173,18 @@ static const std::unordered_set<OutputType> LEGACY_OUTPUT_TYPES {
 };
 
+using WatchOnlySet = std::set<CScript>;
+using WatchKeyMap = std::map<CKeyID, CPubKey>;
```
Done

rkrux reviewed (approved):
ACK 8e9de762f66f3e99e02944006999fe5afb1896e4

Thanks for addressing comments.

```zsh
git range-diff 93c3729...8e9de76
```

adyshimony commented:
ACK 3c4f53d5e1ea30d9a2a4bd1d34501ce419d2aebf

I agree this is a clearer and safer approach.
Reviewed and tested.

Tests:
/build/bin/test_bitcoin
/build/test/functional/test_runner.py wallet_createwallet.py wallet_encryption.py wallet_descriptor.py wallet_importdescriptors.py wallet_signer.py wallet_blank.py wallet_keypool_topup.py

polespinasa commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -603,9 +603,8 @@ std::optional<MigrationData> LegacyDataSPKM::MigrateToDescriptor()
-        auto desc_spk_man = std::make_unique<DescriptorScriptPubKeyMan>(m_storage, w_desc, /*keypool_size=*/0);
-        WITH_LOCK(desc_spk_man->cs_desc_man, desc_spk_man->AddDescriptorKeyWithDB(batch, key, key.GetPubKey()));
-        desc_spk_man->TopUpWithDB(batch);
+        keys.keys.emplace(key.GetPubKey().GetID(), key);
```
in 32fb9ffdd1ee0ffbad6a65640ebef2e90b09991e

nit: I know this is not this PR fault but `keys.keys` is not clear and makes this difficult to understand.
Probably `FlatSigningProvider` variable should have another name? In other parts of the code it is called `provider`.

davidgumberg replied:
nit: Given that the offending is being added here, I think it's in scope for this PR, but if you refuse I will still ACK :)

achow101 replied:
The line that creates `keys` is not in this PR.

davidgumberg replied:
Sure, but the confusing line is being added in this PR, and every user of `keys` is being touched here anyways.

achow101 replied:
Fine, done.

polespinasa reviewed (commented):
Approach ACK
This is much more clean and clear than before.

code reviewed lgtm 3c4f53d5e1ea30d9a2a4bd1d34501ce419d2aebf

I just have some questions and left some non-blocking nits before ACKing it.

murchandamus commented on src/wallet/wallet.cpp:
```diff
@@ -3609,6 +3594,17 @@ void CWallet::SetupDescriptorScriptPubKeyMans()
+{
+    LOCK(cs_wallet);
+    // Skip if blank or no privkeys
+    if (!IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER) &&
```
Wouldn’t `WALLET_FLAG_EXTERNAL_SIGNER` indicate that there are no privkeys? So, why don’t we skip on `WALLET_FLAG_EXTERNAL_SIGNER` true?

Is the comment here wrong?

murchandamus reviewed (commented):
Concept ACK

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -422,30 +422,18 @@ std::shared_ptr<CWallet> CreateWallet(WalletContext& context, const std::string&
+    // Unset the blank flag if not specified by the user
+    if (!create_blank) {
+        wallet->UnsetWalletFlag(WALLET_FLAG_BLANK_WALLET);
+    }
```
> Why could the wallet have the flag set and the user not specify it?

The flag can be set without the user having done so because that's how born-encrypted wallets are implemented. We need to tell wallet creation to not generate any descriptors (or keys, back when this was originally implemented) so that the wallet can then be encrypted, and the descriptors generated afterwards. Since creating a wallet with no descriptors in it is also useful to users, this was made into a wallet flag so that users could also explicitly create such a wallet.

> Wouldn’t it make more sense to optionally allow the creation of a new wallet when you import a descriptor instead of putting the onus on the user to know that they need to create a blank wallet first?

Sure, but that's way out of scope.

davidgumberg replied:
I think born encrypted as an overloaded meaning of the blank flag is kind of a dangerous footgun and a bit hard to comprehend, I haven't checked but I think I recall that this was a lot easier to implement the way it is now in an earlier version of the wallet code, but today it would be a pretty small diff that I think is in scope here to fix this: https://github.com/davidgumberg/bitcoin/commit/130af745309e19934388de2ca1c684b687f9f625

achow101 replied:
Pulled in that commit

davidgumberg commented on src/wallet/wallet.cpp:
```diff
@@ -429,30 +429,18 @@ std::shared_ptr<CWallet> CreateWallet(WalletContext& context, const std::string&
+    }
+
     // Encrypt the wallet
-    if (!passphrase.empty() && !(wallet_creation_flags & WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
```
In commit https://github.com/bitcoin/bitcoin/pull/28333/changes/5767a58f5e6001d6a59505cef3c6796db98fdf69 (_wallet: Consolidate generation setup callers into one function_):

Just a note for reviewers: this is removed because it's redundant with the check above:

https://github.com/bitcoin/bitcoin/blob/8a05adc5f81da483fb4e1b3592acb3e274014298/src/wallet/wallet.cpp#L409

polespinasa replied:
I don't think this is a good idea. Someone could call `CreateNew` with is_encrypted=true and WALLET_FLAG_DISABLE_PRIVATE_KEYS set.

I think a safer approach is to add an assertion to ensure it cannot happen.
`assert(!passphrase.empty() || !(wallet_creation_flags & WALLET_FLAG_DISABLE_PRIVATE_KEYS))`

achow101 replied:
`CreateNew` isn't exposed to the world, it's hard for someone to just call it. There's already a check above, before `CreateNew` for this condition, that's why this can be dropped.

davidgumberg commented on src/wallet/wallet.cpp:
```diff
@@ -429,30 +429,18 @@ std::shared_ptr<CWallet> CreateWallet(WalletContext& context, const std::string&
             status = DatabaseStatus::FAILED_ENCRYPT;
             return nullptr;
         }
-        if (!create_blank) {
-            // Unlock the wallet
-            if (!wallet->Unlock(passphrase)) {
-                error = Untranslated("Error: Wallet was encrypted but could not be unlocked");
-                status = DatabaseStatus::FAILED_ENCRYPT;
-                return nullptr;
-            }
-
-            // Set a seed for the wallet
-            {
-                LOCK(wallet->cs_wallet);
-                wallet->SetupDescriptorScriptPubKeyMans();
-            }
-
-            // Relock the wallet
-            wallet->Lock();
-        }
```
In commit https://github.com/bitcoin/bitcoin/pull/28333/changes/5767a58f5e6001d6a59505cef3c6796db98fdf69 (_wallet: Consolidate generation setup callers into one function_):

For reviewers: This can be dropped because all of this is performed in `EncryptWallet()`.

davidgumberg committed:
refactor: wallet: Don't reuse WALLET_BLANK flag for born-encrypted wallets.

With the split between LoadWallet and CreateNew, it's no longer
necessary to utilize the blank flag to prevent the wallet from having
descriptors automatically being generated. Instead, CreateNew can take a
separate parameter to indicate whether the wallet is to be born
encrypted and therefore should not have any keys generated.

davidgumberg commented on src/wallet/wallet.cpp:
```diff
@@ -429,30 +429,18 @@ std::shared_ptr<CWallet> CreateWallet(WalletContext& context, const std::string&
             status = DatabaseStatus::FAILED_ENCRYPT;
             return nullptr;
         }
-        if (!create_blank) {
-            // Unlock the wallet
-            if (!wallet->Unlock(passphrase)) {
-                error = Untranslated("Error: Wallet was encrypted but could not be unlocked");
-                status = DatabaseStatus::FAILED_ENCRYPT;
-                return nullptr;
-            }
-
-            // Set a seed for the wallet
-            {
-                LOCK(wallet->cs_wallet);
-                wallet->SetupDescriptorScriptPubKeyMans();
-            }
-
-            // Relock the wallet
-            wallet->Lock();
```
In https://github.com/bitcoin/bitcoin/pull/28333/changes/5767a58f5e6001d6a59505cef3c6796db98fdf69#r3127213366 (_wallet: Consolidate generation setup callers into one function_):

This commit is almost a strict refactor but there are two small behavior changes:

1. The case where a wallet is encrypted and then fails to unlock no longer has an explicit check / error message in create.
  - This seems unlikely, but maybe as a sanity check we should keep the check that unlocking works.
  - Code smell that no tests break when this behavior is changed.
2. Before the  born-encrypted wallet was unlocked, locked, then unlocked, then setup descriptors, then locked, now the first no-op unlock, lock is deleted.
  - These should be 100% equivalent in effect, but it is different behavior just to be thorough.

`refactor:` prefix maybe?

achow101 replied:
> Code smell that no tests break when this behavior is changed.

Because it is untestable. There's no way to contrive a test where unlocking immediately after locking results in a failure. The only way that this can happen is if on-disk data is immediately corrupted.

I changed the `Unlock` sanity check in `EncryptWallet` to `return false` if it fails to preserve the sanity check. But as with before, this is not testable.

> refactor: prefix maybe?

Done

davidgumberg replied:
> Because it is untestable. There's no way to contrive a test where unlocking immediately after locking results in a failure. The only way that this can happen is if on-disk data is immediately corrupted.

Yeah that makes sense.

> I changed the `Unlock` sanity check in `EncryptWallet` to `return false` if it fails to preserve the sanity check. But as with before, this is not testable.

I think this was missed

achow101 replied:
Oops, done

davidgumberg commented on src/wallet/walletdb.cpp:
```diff
@@ -771,10 +771,8 @@ static DBErrors LoadDescriptorWalletRecords(CWallet* pwallet, DatabaseBatch& bat
 
-        // Prior to doing anything with this spkm, verify ID compatibility
-        if (id != spkm.GetID()) {
+        if (id != desc.id) {
```
In commit https://github.com/bitcoin/bitcoin/pull/28333/changes/0f294cbb6a255e7c7f72985feffb08a6a27f0458 (_wallet: Load everything into DescSPKM on construction_):

Note for other reviewers, this diff could have been done outside of any of the changes in this PR as the `id` was already computed above at deserialization:

https://github.com/bitcoin/bitcoin/blob/8a05adc5f81da483fb4e1b3592acb3e274014298/src/wallet/walletutil.h#L86

davidgumberg committed:
test: wallet: Check that loading wallet with both unencrypted and encrypted keys fails.

davidgumberg commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -821,6 +821,19 @@ bool LegacyDataSPKM::DeleteRecordsWithDB(WalletBatch& batch)
+    m_wallet_descriptor(descriptor)
+{
+    if (!m_map_keys.empty() && !m_map_crypted_keys.empty()) {
+        throw std::runtime_error("Error: Wallet contains both unencrypted and encrypted keys");
```
In commit https://github.com/bitcoin/bitcoin/pull/28333/changes/0f294cbb6a255e7c7f72985feffb08a6a27f0458 (_wallet: Load everything into DescSPKM on construction_)

This needs a test, e.g.: https://github.com/davidgumberg/bitcoin/commit/92e78971142c35684b7a2efbc859ce9a94311469

achow101 replied:
Pulled in the commit

davidgumberg commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -821,6 +821,19 @@ bool LegacyDataSPKM::DeleteRecordsWithDB(WalletBatch& batch)
+    if (!m_map_keys.empty() && !m_map_crypted_keys.empty()) {
+        throw std::runtime_error("Error: Wallet contains both unencrypted and encrypted keys");
+    }
+    SetCache(m_wallet_descriptor.cache);
```
Maybe to ~enforce that callers are actually setting up the cache add the cache as an argument to the constructor? 

<details>

<summary>

#### diff 

</summary>

```diff
diff --git a/src/wallet/external_signer_scriptpubkeyman.h b/src/wallet/external_signer_scriptpubkeyman.h
index c0bdb037d8..6c181500d3 100644
--- a/src/wallet/external_signer_scriptpubkeyman.h
+++ b/src/wallet/external_signer_scriptpubkeyman.h
@@ -16,8 +16,8 @@ namespace wallet {
 class ExternalSignerScriptPubKeyMan : public DescriptorScriptPubKeyMan
 {
   public:
-  ExternalSignerScriptPubKeyMan(WalletStorage& storage, const uint256& id, WalletDescriptor& descriptor, int64_t keypool_size, const KeyMap& keys, const CryptedKeyMap& ckeys)
-      :   DescriptorScriptPubKeyMan(storage, id, descriptor, keypool_size, keys, ckeys)
+  ExternalSignerScriptPubKeyMan(WalletStorage& storage, const uint256& id, WalletDescriptor& descriptor, int64_t keypool_size, const KeyMap& keys, const CryptedKeyMap& ckeys, const DescriptorCache& cache)
+      :   DescriptorScriptPubKeyMan(storage, id, descriptor, keypool_size, keys, ckeys, cache)
       {}
   ExternalSignerScriptPubKeyMan(WalletStorage& storage, int64_t keypool_size)
       :   DescriptorScriptPubKeyMan(storage, keypool_size)
diff --git a/src/wallet/scriptpubkeyman.cpp b/src/wallet/scriptpubkeyman.cpp
index cf0d0033e5..a49cc99f1d 100644
--- a/src/wallet/scriptpubkeyman.cpp
+++ b/src/wallet/scriptpubkeyman.cpp
@@ -821,7 +821,7 @@ bool LegacyDataSPKM::DeleteRecordsWithDB(WalletBatch& batch)
     return batch.EraseRecords(DBKeys::LEGACY_TYPES);
 }
 
-DescriptorScriptPubKeyMan::DescriptorScriptPubKeyMan(WalletStorage& storage, const uint256& id, WalletDescriptor& descriptor, int64_t keypool_size, const KeyMap& keys, const CryptedKeyMap& ckeys)
+DescriptorScriptPubKeyMan::DescriptorScriptPubKeyMan(WalletStorage& storage, const uint256& id, WalletDescriptor& descriptor, int64_t keypool_size, const KeyMap& keys, const CryptedKeyMap& ckeys, const DescriptorCache& cache)
     : ScriptPubKeyMan(storage),
     m_map_keys(keys),
     m_map_crypted_keys(ckeys),
diff --git a/src/wallet/scriptpubkeyman.h b/src/wallet/scriptpubkeyman.h
index 096e4a57b4..42be4d8f39 100644
--- a/src/wallet/scriptpubkeyman.h
+++ b/src/wallet/scriptpubkeyman.h
@@ -330,7 +330,7 @@ public:
             m_wallet_descriptor(descriptor)
         {}
     //! Create a DescriptorScriptPubKeyMan from existing data (i.e. during loading)
-    DescriptorScriptPubKeyMan(WalletStorage& storage, const uint256& id, WalletDescriptor& descriptor, int64_t keypool_size, const KeyMap& keys, const CryptedKeyMap& ckeys);
+    DescriptorScriptPubKeyMan(WalletStorage& storage, const uint256& id, WalletDescriptor& descriptor, int64_t keypool_size, const KeyMap& keys, const CryptedKeyMap& ckeys, const DescriptorCache& cache);
     DescriptorScriptPubKeyMan(WalletStorage& storage, int64_t keypool_size)
         :   ScriptPubKeyMan(storage),
             m_keypool_size(keypool_size)
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 2eb3242835..3b02856cbb 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -3546,13 +3546,13 @@ void CWallet::ConnectScriptPubKeyManNotifiers()
     }
 }
 
-void CWallet::LoadDescriptorScriptPubKeyMan(uint256 id, WalletDescriptor& desc, const KeyMap& keys, const CryptedKeyMap& ckeys)
+void CWallet::LoadDescriptorScriptPubKeyMan(uint256 id, WalletDescriptor& desc, const KeyMap& keys, const CryptedKeyMap& ckeys, const DescriptorCache& cache)
 {
     DescriptorScriptPubKeyMan* spk_manager;
     if (IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER)) {
-        spk_manager = new ExternalSignerScriptPubKeyMan(*this, id, desc, m_keypool_size, keys, ckeys);
+        spk_manager = new ExternalSignerScriptPubKeyMan(*this, id, desc, m_keypool_size, keys, ckeys, cache);
     } else {
-        spk_manager = new DescriptorScriptPubKeyMan(*this, id, desc, m_keypool_size, keys, ckeys);
+        spk_manager = new DescriptorScriptPubKeyMan(*this, id, desc, m_keypool_size, keys, ckeys, cache);
     }
     AddScriptPubKeyMan(id, std::unique_ptr<ScriptPubKeyMan>(spk_manager));
 }
diff --git a/src/wallet/wallet.h b/src/wallet/wallet.h
index df2659739e..7dec6d72dd 100644
--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -1004,7 +1004,7 @@ public:
     void ConnectScriptPubKeyManNotifiers();
 
     //! Instantiate a descriptor ScriptPubKeyMan from the WalletDescriptor and load it
-    void LoadDescriptorScriptPubKeyMan(uint256 id, WalletDescriptor& desc, const KeyMap& keys, const CryptedKeyMap& ckeys);
+    void LoadDescriptorScriptPubKeyMan(uint256 id, WalletDescriptor& desc, const KeyMap& keys, const CryptedKeyMap& ckeys, const DescriptorCache& cache);
 
     //! Adds the active ScriptPubKeyMan for the specified type and internal. Writes it to the wallet file
     //! @param[in] id The unique id for the ScriptPubKeyMan
diff --git a/src/wallet/walletdb.cpp b/src/wallet/walletdb.cpp
index b9ca9214e2..ca15df0043 100644
--- a/src/wallet/walletdb.cpp
+++ b/src/wallet/walletdb.cpp
@@ -831,9 +831,6 @@ static DBErrors LoadDescriptorWalletRecords(CWallet* pwallet, DatabaseBatch& bat
         });
         result = std::max(result, lh_cache_res.m_result);
 
-        // Set the cache to the WalletDescriptor
-        desc.cache = cache;
-
         // Get unencrypted keys
         KeyMap keys;
         prefix = PrefixStream(DBKeys::WALLETDESCRIPTORKEY, id);
@@ -901,7 +898,7 @@ static DBErrors LoadDescriptorWalletRecords(CWallet* pwallet, DatabaseBatch& bat
         num_ckeys = ckey_res.m_records;
 
         try {
-            pwallet->LoadDescriptorScriptPubKeyMan(id, desc, keys, ckeys);
+            pwallet->LoadDescriptorScriptPubKeyMan(id, desc, keys, ckeys, cache);
         } catch (std::runtime_error& e) {
             strErr = e.what();
             return DBErrors::CORRUPT;
```

</details>

Alternatively... drop the argument from `SetCache()` since this is it's only caller and it uses DSPKM state anyways:

<details>

<summary>

#### diff 

</summary>

```diff
diff --git a/src/wallet/scriptpubkeyman.cpp b/src/wallet/scriptpubkeyman.cpp
index cf0d0033e5..67bdb5597d 100644
--- a/src/wallet/scriptpubkeyman.cpp
+++ b/src/wallet/scriptpubkeyman.cpp
@@ -831,7 +831,7 @@ DescriptorScriptPubKeyMan::DescriptorScriptPubKeyMan(WalletStorage& storage, con
     if (!m_map_keys.empty() && !m_map_crypted_keys.empty()) {
         throw std::runtime_error("Error: Wallet contains both unencrypted and encrypted keys");
     }
-    SetCache(m_wallet_descriptor.cache);
+    SetCache();
 }
 
 util::Result<CTxDestination> DescriptorScriptPubKeyMan::GetNewDestination(const OutputType type)
@@ -1439,11 +1439,10 @@ uint256 DescriptorScriptPubKeyMan::GetID() const
     return m_wallet_descriptor.id;
 }
 
-void DescriptorScriptPubKeyMan::SetCache(const DescriptorCache& cache)
+void DescriptorScriptPubKeyMan::SetCache()
 {
     LOCK(cs_desc_man);
     std::set<CScript> new_spks;
-    m_wallet_descriptor.cache = cache;
     for (int32_t i = m_wallet_descriptor.range_start; i < m_wallet_descriptor.range_end; ++i) {
         FlatSigningProvider out_keys;
         std::vector<CScript> scripts_temp;
diff --git a/src/wallet/scriptpubkeyman.h b/src/wallet/scriptpubkeyman.h
index 096e4a57b4..8cc99310bf 100644
--- a/src/wallet/scriptpubkeyman.h
+++ b/src/wallet/scriptpubkeyman.h
@@ -314,7 +314,7 @@ private:
     // Fetch the SigningProvider for a given index and optionally include private keys. Called by the above functions.
     std::unique_ptr<FlatSigningProvider> GetSigningProvider(int32_t index, bool include_private = false) const EXCLUSIVE_LOCKS_REQUIRED(cs_desc_man);
 
-    void SetCache(const DescriptorCache& cache);
+    void SetCache();
 
 protected:
     WalletDescriptor m_wallet_descriptor GUARDED_BY(cs_desc_man);
```

</details>

achow101 replied:
Taken the second suggestion

davidgumberg commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -70,6 +70,11 @@ static std::optional<std::pair<WalletDescriptor, FlatSigningProvider>> CreateWal
     std::vector<std::unique_ptr<Descriptor>> parsed_descs = Parse(desc_str.value(), keys, error, false);
     if (parsed_descs.empty()) return std::nullopt;
```
In https://github.com/bitcoin/bitcoin/pull/28333/changes/f398d809382631b86f21c7cd793fb7b6362dbdcf (_fuzz: Skip adding descriptor to wallet if it cannot be expanded_):

Context for reviewers for why this commit is here:

https://github.com/bitcoin/bitcoin/pull/28333#issuecomment-1827846220
https://github.com/bitcoin/bitcoin/pull/28333#issuecomment-2035728966

In the next commit, https://github.com/bitcoin/bitcoin/pull/28333/changes/94e0aa68d0bd1e2cc88e3fe9c05a7dac81b9f285 (_wallet: include keys when constructing ,DescriptorSPKM during import_) the DSPKM constructor used by `AddWalletDescriptor()` will now call `UpdateWithSigningProvider`:

https://github.com/bitcoin/bitcoin/blob/94e0aa68d0bd1e2cc88e3fe9c05a7dac81b9f285/src/wallet/scriptpubkeyman.cpp#L821-L829

which throws when top up fails:

https://github.com/bitcoin/bitcoin/blob/94e0aa68d0bd1e2cc88e3fe9c05a7dac81b9f285/src/wallet/scriptpubkeyman.cpp#L1608-L1611

davidgumberg commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -821,6 +818,25 @@ bool LegacyDataSPKM::DeleteRecordsWithDB(WalletBatch& batch)
+    LOCK(cs_desc_man);
+    UpdateWithSigningProvider(batch, provider);
+}
+
```
pure-style-take-if-you-like-otherwise-feel-free-to-disregard-nit:

Maybe one constructor that takes a `std::optional<WalletBatch>` instead of two that differ by one line.

achow101 replied:
Disregarding :)

davidgumberg committed:
test: wallet: Constructing a DSPKM that can't TopUp() throws.

davidgumberg commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1574,10 +1590,27 @@ util::Result<void> DescriptorScriptPubKeyMan::UpdateWalletDescriptor(WalletDescr
+
+    // Top up key pool, to generate scriptPubKeys
+    if (!TopUpWithDB(batch)) {
+        throw std::runtime_error("Could not top up scriptPubKeys");
```
Although unreachable today, since this is part of the DSPKM API it should have a test, maybe e.g.:

https://github.com/davidgumberg/bitcoin/commit/ec12b6b61f5b8615b2d2d86a05c23f01e3af5740

<details>

<summary>diff</summary>

```diff
diff --git a/src/wallet/test/scriptpubkeyman_tests.cpp b/src/wallet/test/scriptpubkeyman_tests.cpp
index 9c62436d6c..b3237b92da 100644
--- a/src/wallet/test/scriptpubkeyman_tests.cpp
+++ b/src/wallet/test/scriptpubkeyman_tests.cpp
@@ -4,6 +4,7 @@
 
 #include <key.h>
 #include <key_io.h>
+#include <test/util/common.h>
 #include <test/util/setup_common.h>
 #include <script/solver.h>
 #include <wallet/scriptpubkeyman.h>
@@ -37,5 +38,17 @@ BOOST_AUTO_TEST_CASE(DescriptorScriptPubKeyManTests)
     BOOST_CHECK(signprov_keypath_nums_h == nullptr);
 }
 
+BOOST_AUTO_TEST_CASE(DescriptorSPKMTopUpFailureThrows)
+{
+    // Attempting to construct a DescriptorSPKM that cannot be topped up should throw.
+    CExtKey extkey;
+    extkey.SetSeed(std::array<std::byte, 32>{});
+    CWallet keystore(m_node.chain.get(), "", CreateMockableWalletDatabase());
+    BOOST_CHECK_EXCEPTION(
+        CreateDescriptor(keystore, "wpkh(" + EncodeExtPubKey(extkey.Neuter()) +
+"/*h)", /*success=*/true),
+        std::runtime_error, HasReason("Could not top up scriptPubKeys"));
+}
+
 BOOST_AUTO_TEST_SUITE_END()
 } // namespace wallet
```

</details>

achow101 replied:
Taken

davidgumberg commented:
crACK https://github.com/bitcoin/bitcoin/commit/6eb2db367785eacd8ed250a854d66c48544ce5d8

This PR RAII-ify's DSPKM's so that they're ready to use after construction, it's a nice cleanup.

One small nit remains: https://github.com/bitcoin/bitcoin/pull/28333#discussion_r3135003098

rkrux commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1426,11 +1439,10 @@ uint256 DescriptorScriptPubKeyMan::GetID() const
 }
 
-void DescriptorScriptPubKeyMan::SetCache(const DescriptorCache& cache)
+void DescriptorScriptPubKeyMan::SetCache()
```
In f4894a8b5cf133ef1958348c820b8f6b1fb37ad5 "wallet: Load everything into DescSPKM on construction"

```diff
diff --git a/src/wallet/scriptpubkeyman.cpp b/src/wallet/scriptpubkeyman.cpp
index 9fc5a9d1cf..9b43c892fd 100644
--- a/src/wallet/scriptpubkeyman.cpp
+++ b/src/wallet/scriptpubkeyman.cpp
@@ -847,7 +847,7 @@ DescriptorScriptPubKeyMan::DescriptorScriptPubKeyMan(WalletStorage& storage, con
     if (!m_map_keys.empty() && !m_map_crypted_keys.empty()) {
         throw std::runtime_error("Error: Wallet contains both unencrypted and encrypted keys");
     }
-    SetCache();
+    Load();
 }
 
 DescriptorScriptPubKeyMan::DescriptorScriptPubKeyMan(WalletStorage& storage, WalletBatch& batch, int64_t keypool_size, const CExtKey& master_key, OutputType addr_type, bool internal)
@@ -1461,7 +1461,7 @@ uint256 DescriptorScriptPubKeyMan::GetID() const
     return m_wallet_descriptor.id;
 }
 
-void DescriptorScriptPubKeyMan::SetCache()
+void DescriptorScriptPubKeyMan::Load()
 {
     LOCK(cs_desc_man);
     std::set<CScript> new_spks;
diff --git a/src/wallet/scriptpubkeyman.h b/src/wallet/scriptpubkeyman.h
index 15f8179b45..e0dc58f020 100644
--- a/src/wallet/scriptpubkeyman.h
+++ b/src/wallet/scriptpubkeyman.h
@@ -314,7 +314,7 @@ private:
     // Fetch the SigningProvider for a given index and optionally include private keys. Called by the above functions.
     std::unique_ptr<FlatSigningProvider> GetSigningProvider(int32_t index, bool include_private = false) const EXCLUSIVE_LOCKS_REQUIRED(cs_desc_man);
 
-    void SetCache();
+    void Load();
 
     void AddDescriptorKey(const CKey& key, const CPubKey &pubkey);
     void UpdateWithSigningProvider(WalletBatch& batch, const FlatSigningProvider& signing_provider) EXCLUSIVE_LOCKS_REQUIRED(cs_desc_man);

```

The name `SetCache` for this method seems outdated now because neither does it accept nor does it set the (descriptor) cache anymore, and it only sets the private members of `DescriptorScriptPubKeyMan`. Though it can be argued that these internal members of the class are sort of a cache but I am not inclined towards that because the term cache is usually reserved for something that's explicitly designed for quicker and frequent access and using cache for most private members (if not all) seems like a stretch.  

More importantly, this method is called from only one constructor that's called during loading the descriptor records from the wallet database, so I believe a simple `Load()` would be more pertinent.

I do notice that there is a wallet `TopUpCallback` method call at the end of this method that signals to me that this method is doing more than what's necessary and therefore believe that call could be moved one level up in the constructor calling this method but that would be unrelated changes that needn't be done in this PR.

achow101 replied:
Done

rkrux commented on src/wallet/wallet.cpp:
```diff
@@ -896,12 +861,13 @@ bool CWallet::EncryptWallet(const SecureString& strWalletPassphrase)
+            return false;
         }
+
+        // Generate new descriptors or seed if not blank or disable private keys
```
In aac8533725112e0330da87f32b63386a369f8a9a "wallet: Consolidate generation setup callers into one function"

> if not blank or disable private keys

This phrase in the comment is leaking the internals of the below method here, which seems unnecessary at the least and confusing at the most. Remove it?

```diff
- // Generate new descriptors or seed if not blank or disable private keys
```

achow101 replied:
Removed

rkrux commented on src/wallet/wallet.cpp:
```diff
@@ -385,18 +385,12 @@ std::shared_ptr<CWallet> CreateWallet(WalletContext& context, const std::string&
 
     uint64_t wallet_creation_flags = options.create_flags;
     const SecureString& passphrase = options.create_passphrase;
+    bool is_encrypted = !passphrase.empty();
```
In 03993ad4756d4d4e3058a11428204ed1e630431f "refactor: wallet: Don't reuse WALLET_BLANK flag for born-encrypted wallets."

Make the name more explicit? The name `is_encrypted` (not this variable) can be `true` for also the wallets that are encrypted post creation.
 
```diff
- bool is_encrypted = !passphrase.empty();
+ bool born_encrypted = !passphrase.empty(); 
```

achow101 replied:
Renamed

rkrux commented on src/wallet/test/util.cpp:
In 03993ad "refactor: wallet: Don't reuse WALLET_BLANK flag for born-encrypted wallets."

Comment about the commit - Seems fine at first glance to not use the flag but the message doesn't mention the reason.

achow101 replied:
Expanded the commit message

rkrux commented on src/wallet/wallet.cpp:
```diff
@@ -3090,7 +3079,10 @@ std::shared_ptr<CWallet> CWallet::CreateNew(WalletContext& context, const std::s
         assert(walletInstance->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS));
 
-        walletInstance->SetupWalletGeneration();
+        // Born encrypted wallets need to be created blank first.
```
In 03993ad4756d4d4e3058a11428204ed1e630431f "refactor: wallet: Don't reuse WALLET_BLANK flag for born-encrypted wallets."

> blank first.

Now that the usage of blank flag is removed in born encrypted wallets, consider removing the blank terminology as well.

```diff
- // Born encrypted wallets need to be created blank first.
+ // Born encrypted wallets will have their keys generated later.
```

achow101 replied:
Done

rkrux commented on src/wallet/wallet.cpp:
In 03993ad "refactor: wallet: Don't reuse WALLET_BLANK flag for born-encrypted wallets."

This needs to be inside the is_encrypted conditional as well, I suppose. Seems like a no-op for encrypted wallet creation flow.

achow101 replied:
It does not need to be, it's a no-op when the wallet has no descriptors. It's a no-op for blank wallets too.

rkrux reviewed (commented):
Paused review at f4894a8b5cf133ef1958348c820b8f6b1fb37ad5.

Nice to see more activity on this PR, moving it closer to merge.

However, the PR seems to have changed a bit from last year when I reviewed it and I'm unable to place the second commit in the context of this PR [refactor: wallet: Don't reuse WALLET_BLANK flag for born-encrypted wallets.](https://github.com/bitcoin/bitcoin/pull/28333/changes/03993ad4756d4d4e3058a11428204ed1e630431f) now - it seems fine to do but going through it seems like a deviation while reviewing the PR that's supposed to construct SPKMs with all data rather than loaded progressively.

I see the second commit was added recently (https://github.com/bitcoin/bitcoin/pull/28333#discussion_r3134291830) but I really feel it should be its own PR. I ended up adding suggestions in this commit but I don't see a reason why they should be addressed in this PR that's supposed to do something else.

polespinasa commented on src/wallet/wallet.cpp:
```diff
@@ -3695,6 +3657,19 @@ void CWallet::SetupDescriptorScriptPubKeyMans()
+    // We only want to call Setup if we are not blank,
+    // have private keys (not having private keys implies blank),
+    // or need to get descriptors from an external signer.
+    if (!IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER) &&
```
In aac8533725112e0330da87f32b63386a369f8a9a _wallet: Consolidate generation setup callers into one function_

This comment doesn’t exactly reflect the condition. The logic skips setup for non-external signers when the wallet is blank or has private keys disabled. Suggest rewording the comment to match the code behavior.

```
// Skip setup for non-external-signer wallets that are either blank
// or have private keys disabled (not having private keys implies blank).
```

achow101 replied:
Done

polespinasa commented on src/wallet/wallet.h:
```diff
@@ -1039,6 +1033,9 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     void SetupOwnDescriptorScriptPubKeyMans(WalletBatch& batch) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
 
+    //! Setup new descriptors or seed for new address generation
+    void SetupWalletGeneration();
```
in aac8533 wallet: Consolidate generation setup callers into one function
Missing `EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)` ?

achow101 replied:
This function was written as acquiring the lock, so `EXCLUSIVE_LOCKS_REQUIRED` is incorrect for that. However, since all callers take the lock beforehand, I changed it to use `AssertLockHeld` which does require the annotation.

polespinasa commented on src/wallet/wallet.cpp:
```diff
@@ -385,18 +385,12 @@ std::shared_ptr<CWallet> CreateWallet(WalletContext& context, const std::string&
 
     uint64_t wallet_creation_flags = options.create_flags;
     const SecureString& passphrase = options.create_passphrase;
+    bool is_encrypted = !passphrase.empty();
```
In 03993ad4756d4d4e3058a11428204ed1e630431f _refactor: wallet: Don't reuse WALLET_BLANK flag for born-encrypted wallets._

feel-free-to-ignore-nit: I suggest use another name like `should_encrypt` as at creation time wallet is not encrypted yet, it will be.

achow101 replied:
Renamed to `born_encrypted`

polespinasa commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -821,6 +821,19 @@ bool LegacyDataSPKM::DeleteRecordsWithDB(WalletBatch& batch)
     return batch.EraseRecords(DBKeys::LEGACY_TYPES);
 }
 
+DescriptorScriptPubKeyMan::DescriptorScriptPubKeyMan(WalletStorage& storage, const uint256& id, WalletDescriptor& descriptor, int64_t keypool_size, const KeyMap& keys, const CryptedKeyMap& ckeys)
```
in f4894a8b5cf133ef1958348c820b8f6b1fb37ad5 _wallet: Load everything into DescSPKM on construction_

`id` is never used, can be dropped here and in `ExternalSignerScriptPubKeyMan`

achow101 replied:
Dropped

polespinasa commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -821,6 +821,19 @@ bool LegacyDataSPKM::DeleteRecordsWithDB(WalletBatch& batch)
+    m_wallet_descriptor(descriptor)
+{
+    if (!m_map_keys.empty() && !m_map_crypted_keys.empty()) {
+        throw std::runtime_error("Error: Wallet contains both unencrypted and encrypted keys");
```
in f4894a8b5cf133ef1958348c820b8f6b1fb37ad5 _wallet: Load everything into DescSPKM on construction_

nit: Seems that we don't add `Error:` prefix in front of runtime errors (have only seen 2-3 examples in wallet code). Probably is better to drop it for consistency.
```bash
$ git grep "std::runtime_error" | grep "Error:"
src/util/subprocess.h:class CalledProcessError: public std::runtime_error
src/util/subprocess.h:class OSError: public std::runtime_error
src/wallet/scriptpubkeyman.cpp:        throw std::runtime_error("Error: Wallet contains both unencrypted and encrypted keys");
src/wallet/scriptpubkeyman.cpp:            throw std::runtime_error("Error: Unable to expand wallet descriptor from cache");
src/wallet/scriptpubkeyman.cpp:                throw std::runtime_error(strprintf("Error: Already loaded script at index %d as being at index %d", i, m_map_script_pub_keys[script]));
src/wallet/wallet.cpp:        })) throw std::runtime_error("Error: cannot process db transaction for descriptors setup");
src/wallet/wallet.cpp:        if (!batch.TxnBegin()) throw std::runtime_error("Error: cannot create db transaction for descriptors import");
src/wallet/wallet.cpp:        if (!batch.TxnCommit()) throw std::runtime_error("Error: cannot commit db transaction for descriptors import");
```

achow101 replied:
Done

polespinasa commented on src/wallet/walletdb.cpp:
```diff
@@ -896,12 +894,19 @@ static DBErrors LoadDescriptorWalletRecords(CWallet* pwallet, DatabaseBatch& bat
             value >> privkey;
 
-            spk_man->AddCryptedKey(pubkey.GetID(), pubkey, privkey);
+            ckeys[pubkey.GetID()] = make_pair(pubkey, privkey);
```
in f4894a8b5cf133ef1958348c820b8f6b1fb37ad5 _wallet: Load everything into DescSPKM on construction_

nit: `ckeys[pubkey.GetID()] = std::make_pair(pubkey, privkey);`

achow101 replied:
Done

polespinasa commented on test/functional/wallet_descriptor.py:
```diff
@@ -268,6 +269,25 @@ def run_test(self):
+        with conn:
+            key_prefix = ser_string(b"walletdescriptorkey")
+            ckey_prefix = ser_string(b"walletdescriptorckey")
+            rows = conn.execute('Select key, value FROM main').fetchall()
```
in 110628b0fee6cb66b3194ef082a3dfa7f9356fbc _test: wallet: Check that loading wallet with both unencrypted and encrypted keys fails._

nit: should be `SELECT`, it's more standard style

achow101 replied:
Done

polespinasa commented on test/functional/wallet_descriptor.py:
```diff
@@ -268,6 +269,25 @@ def run_test(self):
+            k, v = key_rows[0]
+            conn.execute('INSERT INTO main VALUES(?, ?)', (k.replace(key_prefix, ckey_prefix), v))
+        conn.close()
+        with self.nodes[0].assert_debug_log(["Wallet contains both unencrypted and encrypted keys"]):
```
in 110628b0fee6cb66b3194ef082a3dfa7f9356fbc _test: wallet: Check that loading wallet with both unencrypted and encrypted keys fails._

As mentioned in https://github.com/bitcoin/bitcoin/pull/28333/changes/f4894a8b5cf133ef1958348c820b8f6b1fb37ad5#r3148895417

Here the `Error:` prefix is not present. (Should be removed in the constructor)

polespinasa commented on src/wallet/test/fuzz/scriptpubkeyman.cpp:
```diff
@@ -70,6 +70,11 @@ static std::optional<std::pair<WalletDescriptor, FlatSigningProvider>> CreateWal
+    FlatSigningProvider out_keys;
+    std::vector<CScript> scripts_temp;
+    DescriptorCache temp_cache;
+    if (!parsed_descs.at(0)->Expand(0, keys, scripts_temp, out_keys, &temp_cache)) return std::nullopt;
```
in 804dc85f30a692531765eb0a8a9148a3e7afa3ff _fuzz: Skip adding descriptor to wallet if it cannot be expanded_

nit: `out_keys`, `scripts_temp` and `temp_cache` are populated but then discarded which can be a bit confusing, maybe a small comment with something like: `// Only check whether Expand succeeds; discard other outputs` can be helpfull?

achow101 replied:
Added a comment

polespinasa commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1573,10 +1589,27 @@ util::Result<void> DescriptorScriptPubKeyMan::UpdateWalletDescriptor(WalletDescr
+
+    // Top up key pool, to generate scriptPubKeys
+    if (!TopUpWithDB(batch)) {
+        throw std::runtime_error("Could not top up scriptPubKeys");
```
in fcffb36faf90c82c47eb9076e7abbd9fbc70bee6 _wallet: include keys when constructing DescriptorSPKM during import_

I am not really familiar with error handling, but the error type changed here. I think this way it cannot be translated? Which is a problem as this is propagated till `importdescriptor` RPC.

not sure of this tbh feel free to ignore if I am mistaken

achow101 replied:
It doesn't matter for RPCs as the output to RPC is never translated. This could be an issue for the GUI once import descriptors in the GUI is implemented. However, I'm pretty sure that this condition can't be hit in practice as it would mean that the descriptor cannot be expanded even when the private keys are available. This is checked for within `importdescriptors` before we get to this point.

polespinasa replied:
> It doesn't matter for RPCs as the output to RPC is never translated

didn't know it, make sense then :)

polespinasa replied:
Even if it can't be hit, I think it is never catch?

polespinasa commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -324,11 +327,8 @@ class DescriptorScriptPubKeyMan : public ScriptPubKeyMan
-            m_wallet_descriptor(descriptor)
-        {}
+    DescriptorScriptPubKeyMan(WalletStorage& storage, WalletDescriptor& descriptor, int64_t keypool_size, const FlatSigningProvider& provider);
+    DescriptorScriptPubKeyMan(WalletStorage& storage, WalletBatch& batch, WalletDescriptor& descriptor, int64_t keypool_size, const FlatSigningProvider& provider);
```
in fcffb36faf90c82c47eb9076e7abbd9fbc70bee6 _wallet: include keys when constructing DescriptorSPKM during import_

nit: maybe document a bit more the constructors, when to use each?

achow101 replied:
Added comments

polespinasa commented on src/wallet/wallet.cpp:
```diff
@@ -3554,7 +3554,7 @@ void CWallet::LoadDescriptorScriptPubKeyMan(uint256 id, WalletDescriptor& desc,
 {
     AssertLockHeld(cs_wallet);
-    auto spk_manager = std::unique_ptr<DescriptorScriptPubKeyMan>(new DescriptorScriptPubKeyMan(*this, m_keypool_size));
+    auto spk_manager = std::unique_ptr<DescriptorScriptPubKeyMan>(new DescriptorScriptPubKeyMan(*this, batch, m_keypool_size, master_key, output_type, internal));
```
in d946ecf3b1930e94dbcbf3a4470cffa1ec6d0a03 _wallet: Setup new autogenerated descriptors on construction_

In the old code `SetupDescriptorGeneration()` was called after the `HasEncryption() && IsLocked()` check, but now it is called before (inside `DescriptorScriptPubKeyMan()`). This means that the batch already received unencrypted key writes before the encryption check.

achow101 replied:
Moved this down after the encryption check.

Also dropped the unnecessary `CheckDecryptionKeys` and `Encrypt`.

polespinasa commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1161,14 +1168,14 @@ bool DescriptorScriptPubKeyMan::AddDescriptorKeyWithDB(WalletBatch& batch, const
     assert(m_storage.IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS));
 
     // Ignore when there is already a descriptor
     if (m_wallet_descriptor.descriptor) {
```
in d946ecf3b1930e94dbcbf3a4470cffa1ec6d0a03 _wallet: Setup new autogenerated descriptors on construction_

`SetupDescriptorGeneration` is a private function that is only called from a constructor. I think this can never happen because of that. Maybe changing it to assert is better.

achow101 replied:
Done

polespinasa commented on src/wallet/test/scriptpubkeyman_tests.cpp:
```diff
@@ -37,5 +38,17 @@ BOOST_AUTO_TEST_CASE(DescriptorScriptPubKeyManTests)
+    CWallet keystore(m_node.chain.get(), "", CreateMockableWalletDatabase());
+    BOOST_CHECK_EXCEPTION(
+        CreateDescriptor(keystore, "wpkh(" + EncodeExtPubKey(extkey.Neuter()) +
+"/*h)", /*success=*/true),
```
in 83c810a7ae3f3f016c807d4efabe4b3264c6b5ff _test: wallet: Constructing a DSPKM that can't TopUp() throws._

nit: you are breaking a string with a line jump.

achow101 replied:
Fixed

polespinasa reviewed (changes requested):
reviewed 83c810a7ae3f3f016c807d4efabe4b3264c6b5ff

left some comments

pseudoramdom commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -314,22 +314,33 @@ class DescriptorScriptPubKeyMan : public ScriptPubKeyMan
-            m_keypool_size(keypool_size)
-        {}
+    //! Create a new DescriptorScriptPubKeyMan from an existing descriptor (i.e. from an import)
+    DescriptorScriptPubKeyMan(WalletStorage& storage, WalletDescriptor& descriptor, int64_t keypool_size, const FlatSigningProvider& provider);
```
(I’m new here and not very familiar with the project conventions, so feel free to disregard if this does not fit the local style)

f4894a8b5cf133ef1958348c820b8f6b1fb37ad5
These constructor overloads seem to be doing more than just creating a `DescriptorScriptPubKeyMan`. Some of them import keys, write wallet records, top up scripts, or change wallet state. This makes the call sites look like simple object construction even though different overloads have different side effects and failure modes. 

Could these be named factory functions like `CreateForImport`, `CreateGenerated`, or `LoadFromDatabase` make the intent and error handling clearer? 

```cpp
static util::Result<std::unique_ptr<DescriptorScriptPubKeyMan>> CreateForImport(
          WalletStorage& storage,
          WalletDescriptor descriptor,
          int64_t keypool_size,
          const FlatSigningProvider& provider);
```

Alternatively, would it help to just add the named factory wrappers?

achow101 replied:
The constructors are useful for inheriting parent behavior that is useful in children. However, using factory functions does seem like it is easier to read rather than the overloads, so I've reimplemented most of the constructors in that way, and made the remaining be wrappers around the (now private) constructors.

polespinasa commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -166,14 +166,18 @@ static const std::unordered_set<OutputType> LEGACY_OUTPUT_TYPES {
+using KeyMap = std::map<CKeyID, CKey>;
+using CryptedKeyMap = std::map<CKeyID, std::pair<CPubKey, std::vector<unsigned char>>>;
+using ScriptPubKeyMap = std::map<CScript, int32_t>; // Map of scripts to descriptor range index
+using PubKeyMap = std::map<CPubKey, int32_t>; // Map of pubkeys involved in scripts to descriptor range index
```
in fb184af3edd22670d5f1f0c9d4df7dd5240c51a7 _wallet: Load everything into DescSPKM on construction_

The last two definitions are done inside `wallet` namespace but only used inside the class. Maybe can be moved inside it to not pollute the namespace.

rkrux replied:
> but only used inside the class

The first two are used in multiple places.

polespinasa replied:
> > but only used inside the class
> 
> The first two are used in multiple places.

You're right, I fixed the comment

achow101 replied:
Done

polespinasa commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -166,14 +166,18 @@ static const std::unordered_set<OutputType> LEGACY_OUTPUT_TYPES {
 class LegacyDataSPKM : public ScriptPubKeyMan, public FillableSigningProvider
 {
-private:
+protected:
```
in fb184af3edd22670d5f1f0c9d4df7dd5240c51a7 _wallet: Load everything into DescSPKM on construction_

Maybe an explanation on why change this to `protected` in the commit message?

rkrux replied:
This change seems unnecessary because it builds fine without it - most likely outdated diff.

achow101 replied:
Reverted

polespinasa commented on src/wallet/test/scriptpubkeyman_tests.cpp:
```diff
@@ -37,5 +38,16 @@ BOOST_AUTO_TEST_CASE(DescriptorScriptPubKeyManTests)
+    extkey.SetSeed(std::array<std::byte, 32>{});
+    CWallet keystore(m_node.chain.get(), "", CreateMockableWalletDatabase());
+    BOOST_CHECK_EXCEPTION(
+        CreateDescriptor(keystore, "wpkh(" + EncodeExtPubKey(extkey.Neuter()) + "/*h)", /*success=*/true),
```
in dcd91fd710ac10d184b18121f5044f37ca29fa6b _test: wallet: Constructing a DSPKM that can't TopUp() throws._

This is a bit confusing, we have `*success=*/true` but at the same time we expect it to throw an exception.

Also maybe a comment saying why this has to throw would help. I guess is because we require priv keys when using `/*h`.

achow101 replied:
Expanded the comment above.

polespinasa reviewed (commented):
Thanks for addressing comments :)

Just three more things I just saw and I will ack :P

rkrux commented on src/wallet/wallet.h:
```diff
@@ -871,7 +871,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
 
     /* Initializes, creates and returns a new CWallet; returns a null pointer in case of an error */
-    static std::shared_ptr<CWallet> CreateNew(WalletContext& context, const std::string& name, std::unique_ptr<WalletDatabase> database, uint64_t wallet_creation_flags, bilingual_str& error, std::vector<bilingual_str>& warnings);
+    static std::shared_ptr<CWallet> CreateNew(WalletContext& context, const std::string& name, std::unique_ptr<WalletDatabase> database, uint64_t wallet_creation_flags, bool is_encrypted, bilingual_str& error, std::vector<bilingual_str>& warnings);
```
In 2318f5dce1b1349738c5adc0a34dec6fd45938d0 "refactor: wallet: Don't reuse WALLET_BLANK flag for born-encrypted wallets."

To fix the clang tidy error: https://github.com/bitcoin/bitcoin/actions/runs/25028250329/job/73304196483?pr=28333
```diff
-    static std::shared_ptr<CWallet> CreateNew(WalletContext& context, const std::string& name, std::unique_ptr<WalletDatabase> database, uint64_t wallet_creation_flags, bool is_encrypted, bilingual_str& error, std::vector<bilingual_str>& warnings);
+    static std::shared_ptr<CWallet> CreateNew(WalletContext& context, const std::string& name, std::unique_ptr<WalletDatabase> database, uint64_t wallet_creation_flags, bool born_encrypted, bilingual_str& error, std::vector<bilingual_str>& warnings);
```

achow101 replied:
Done

rkrux commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -303,6 +303,13 @@ class DescriptorScriptPubKeyMan : public ScriptPubKeyMan
     mutable std::map<uint256, MuSig2SecNonce> m_musig2_secnonces;
 
+    //! Create a new DescriptorScriptPubKeyMan from an existing descriptor (i.e. from an import)
+    DescriptorScriptPubKeyMan(WalletStorage& storage, WalletDescriptor& descriptor, int64_t keypool_size, const FlatSigningProvider& provider)
```
In b1f82549616b37d1d24747fc28dd2b9872ff57b7 "wallet: include keys when constructing DescriptorSPKM during import"

provider is not needed in this constructor.

```diff
diff --git a/src/wallet/scriptpubkeyman.cpp b/src/wallet/scriptpubkeyman.cpp
index c4400b0c06..473ad2d7d7 100644
--- a/src/wallet/scriptpubkeyman.cpp
+++ b/src/wallet/scriptpubkeyman.cpp
@@ -820,7 +820,7 @@ bool LegacyDataSPKM::DeleteRecordsWithDB(WalletBatch& batch)
 
 std::unique_ptr<DescriptorScriptPubKeyMan> DescriptorScriptPubKeyMan::CreateFromImport(WalletStorage& storage, WalletDescriptor& descriptor, int64_t keypool_size, const FlatSigningProvider& provider)
 {
-    auto spkm = std::unique_ptr<DescriptorScriptPubKeyMan>(new DescriptorScriptPubKeyMan(storage, descriptor, keypool_size, provider));
+    auto spkm = std::unique_ptr<DescriptorScriptPubKeyMan>(new DescriptorScriptPubKeyMan(storage, descriptor, keypool_size));
     LOCK(spkm->cs_desc_man);
     WalletBatch batch(storage.GetDatabase());
     spkm->UpdateWithSigningProvider(batch, provider);
@@ -829,7 +829,7 @@ std::unique_ptr<DescriptorScriptPubKeyMan> DescriptorScriptPubKeyMan::CreateFrom
 
 std::unique_ptr<DescriptorScriptPubKeyMan> DescriptorScriptPubKeyMan::CreateFromMigration(WalletStorage& storage, WalletBatch& batch, WalletDescriptor& descriptor, int64_t keypool_size, const FlatSigningProvider& provider)
 {
-    auto spkm = std::unique_ptr<DescriptorScriptPubKeyMan>(new DescriptorScriptPubKeyMan(storage, descriptor, keypool_size, provider));
+    auto spkm = std::unique_ptr<DescriptorScriptPubKeyMan>(new DescriptorScriptPubKeyMan(storage, descriptor, keypool_size));
     LOCK(spkm->cs_desc_man);
     spkm->UpdateWithSigningProvider(batch, provider);
     return spkm;
diff --git a/src/wallet/scriptpubkeyman.h b/src/wallet/scriptpubkeyman.h
index 3cae66b2e8..621630c948 100644
--- a/src/wallet/scriptpubkeyman.h
+++ b/src/wallet/scriptpubkeyman.h
@@ -304,7 +304,7 @@ private:
     mutable std::map<uint256, MuSig2SecNonce> m_musig2_secnonces;
 
     //! Create a new DescriptorScriptPubKeyMan from an existing descriptor (i.e. from an import)
-    DescriptorScriptPubKeyMan(WalletStorage& storage, WalletDescriptor& descriptor, int64_t keypool_size, const FlatSigningProvider& provider)
+    DescriptorScriptPubKeyMan(WalletStorage& storage, WalletDescriptor& descriptor, int64_t keypool_size)
         : ScriptPubKeyMan(storage),
         m_keypool_size(keypool_size),
         m_wallet_descriptor(descriptor)

```

achow101 replied:
Done

w0xlt commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1579,10 +1599,27 @@ util::Result<void> DescriptorScriptPubKeyMan::UpdateWalletDescriptor(WalletDescr
+    // Add the private keys to the descriptor
+    for (const auto& entry : signing_provider.keys) {
+        const CKey& key = entry.second;
+        AddDescriptorKeyWithDB(batch, key, key.GetPubKey());
```
If the result of `AddDescriptorKeyWithDB` is ignored, the operation could look successful while the wallet had not actually persisted the supplied private key.

```suggestion
        if (!AddDescriptorKeyWithDB(batch, key, key.GetPubKey())) {
                throw std::runtime_error(std::string(__func__) + ": writing descriptor private key failed");
        }
```
Test for this:

```diff
diff --git a/src/wallet/test/scriptpubkeyman_tests.cpp b/src/wallet/test/scriptpubkeyman_tests.cpp
index 74389a5496..d1b63d0b02 100644
--- a/src/wallet/test/scriptpubkeyman_tests.cpp
+++ b/src/wallet/test/scriptpubkeyman_tests.cpp
@@ -50,5 +50,26 @@ BOOST_AUTO_TEST_CASE(desc_spkm_topup_fail)
         std::runtime_error, HasReason("Could not top up scriptPubKeys"));
 }
 
+BOOST_AUTO_TEST_CASE(desc_spkm_privkey_write_fail)
+{
+    CWallet keystore(m_node.chain.get(), "", CreateMockableWalletDatabase());
+    keystore.SetWalletFlag(WALLET_FLAG_DESCRIPTORS);
+    BOOST_REQUIRE(keystore.EncryptWallet(SecureString{"passphrase"}));
+    BOOST_REQUIRE(keystore.IsLocked());
+
+    const auto key = GenerateRandomKey();
+    FlatSigningProvider keys;
+    std::string error;
+    auto parsed_descs = Parse("wpkh(" + EncodeSecret(key) + ")", keys, error, /*require_checksum=*/false);
+    BOOST_REQUIRE_EQUAL(parsed_descs.size(), 1);
+
+    WalletDescriptor w_desc(std::move(parsed_descs.at(0)), /*creation_time=*/1, /*range_start=*/0, /*range_end=*/1, /*next_index=*/0);
+
+    LOCK(keystore.cs_wallet);
+    BOOST_CHECK_EXCEPTION(
+        keystore.AddWalletDescriptor(w_desc, keys, /*label=*/"", /*internal=*/false),
+        std::runtime_error, HasReason("writing descriptor private key failed"));
+}
+
 BOOST_AUTO_TEST_SUITE_END()
 } // namespace wallet
```

achow101 replied:
Fixed. I don't think a test is necessary.

achow101 committed:
wallet migration, fuzz: Migrate hd seed once

If a wallet has multiple HD chains that have the same seed, we should
only migrate that seed a single time.

This fixes a fuzz crash that occurs once the return value of
AddDescriptorKeyWithDB is checked during descriptor construction.

achow101 commented:
Added a commit to fix the fuzz crash.

polespinasa commented:
ACK 451fdd26a4f638e4f68f076ba2a891222cea380c

thanks for addressing all comments, the code lgtm :)

w0xlt reviewed (commented):
ACK 451fdd26a4f638e4f68f076ba2a891222cea380c

davidgumberg commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -821,6 +818,23 @@ bool LegacyDataSPKM::DeleteRecordsWithDB(WalletBatch& batch)
+    return spkm;
+}
+
+std::unique_ptr<DescriptorScriptPubKeyMan> DescriptorScriptPubKeyMan::CreateFromMigration(WalletStorage& storage, WalletBatch& batch, WalletDescriptor& descriptor, int64_t keypool_size, const FlatSigningProvider& provider)
```
In https://github.com/bitcoin/bitcoin/pull/28333/changes/aa4f7823aa18eea7e310a355981bbf70c0496b98 (_wallet: include keys when constructing DescriptorSPKM during import_)

disregard-worthy-nit: the names `CreateFromImport()` and `CreateFromMigration()` seem to be more about the callers than telling you anything about the difference between these two functions.

achow101 replied:
Couldn't think of better names :woman_shrugging:

davidgumberg commented:
re crACK https://github.com/bitcoin/bitcoin/commit/451fdd26a4f638e4f68f076ba2a891222cea380c

I reviewed the range-diff, the main changes since my last review:

- fuzz: Migrate hd seed added.
- Factory pattern suggested by @pseudoramdom 
- Thread annotation suggested by @polespinasa

fanquake commented:
Possible that this change has resulted in the fuzz crash in #35434.

kevkevinpal commented:
> Possible that this change has resulted in the fuzz crash in #35434.

It looks like specifically on aa4f7823aa18eea7e310a355981bbf70c0496b98 it crashes

to reproduce the crash
```
# Add input
echo 'IQIAJ1sAo6MYAcBQGP+VdP////////83v9sAAAAAAAAAAAAAAP8VAAAAAAAAAUUAAAAAWwD+////////CwAAAAAAAAB///8AAAAAAAAAAAAAAP8VAAAAAAAAAUUAAAAAAAD/k5OTk5OTk5OTk5OTk5OTk5OTk5OTk5OTk5OTk5MhAgAnWwCjoxgDwFAY/5V0/////////z+/2wAAAAABAAAAAAAA/xUAAAAAAAABRQAAAAAAAP////////8LAAAAAAAA/////wAA/P8AAAAAAAAA/xUAAAAAAAABRQAAAAAAAP+Tk5OTk5OTk5OTk5OTk5OTk5OTk5OTk5OTk5OTkyECACdbAKOjGAPqUBhRa3Q+Pj4+Pj4+Pj4+Pj4+Pj4+Pj48Pj4+Pj4+wsHBuT4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+/////////zm/2wAAAAAAAAAAhqurq6urAAD/FQAAXAAAANoAAAAAAAAAAAAAAPgA0wAAAAAAAAAA/v/DX+Wel3+Tk5OTk5OTk5OTk5OTk5OTk0aB6YU2k5OTk5OTk5OTk5MhAgAAAAAAAP+Tk5OTk5OTk5OTk5OTk5OTk5OTopOTk5OTk5OTkyECACdbAKOjGAPAUBj/lXT/////////P7/bAAAAAAAAAAAAAAAAAABkAAAAAAFFAAAAAAAA/////////wsAAAAAAAD/////AAD8/wAAAAAAAAD/FQAAAAAAAAFFAAAAAAAA/5OTk5OTk5OTk5OTk5OTk5OTk5OTk5OTk5OTk5OTIQIAJ1sAo6MYA+pQGFFrdEA+Pj4+Pj4+Pj4+Pj4+Pj4+Pjw+Pj4+Pj7CwcG5Pj4+Pj4+Pv83v9sAAAAAAAAAAAAAAP8VAAAAAAAAAUUAAAAAWwD+/////v//CwAAAAAAAAB///8AAAAAAAAAAAAAAP8VAAAAAAAAAUUAAAAAAAD/k5OTk5OTk5OTk5OTk5N6k5OTk5OTk5OTk5OTk5MhAgAnWwCjoxgDwFAY/5V0/////////z+/2wAAAAAAAAAAAAAA/xUAAAAAAAABRQAAAAAAAP////////8LAAAAAAAA/////wAA/P8AAAAAAAAA/xUAAAAAAAABRQAAAAAAAP+Tk5OTk5OTk5OTk5OTk5OTk5OTk5OTk5OTk5OTkyECACdbAKOjGAPqUBhRa3Q+Pj4+Pj4+Pj4+Pj4+Pj4+Pj48Pj4+Pj4+wsHBuT4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+/////////zm/2wAAAAAAAAAAhqurq6urAAD/Fd7/WwAAANoAAAAAAAAAAAAAAPgA0wAAAAAAAAAA/v/YX+Wel3+Tk5MAAP+Y/6Ghnpd/k5OTk5OTkwAAAKE='   | base64 -d > spkmmigrate_input

# Checkout failing commit
git checkout aa4f7823aa18eea7e310a355981bbf70c0496b98
cmake --preset=libfuzzer && cmake --build build_fuzz --parallel

# Run failing input
FUZZ=spkm_migration build_fuzz/bin/fuzz -runs=1 spkmmigrate_input

# Test with working/prior commit
git checkout aa4f7823aa18eea7e310a355981bbf70c0496b98^
cmake --preset=libfuzzer && cmake --build build_fuzz --parallel

# Run failing input (observe it passes)
FUZZ=spkm_migration build_fuzz/bin/fuzz -runs=1 spkmmigrate_input
```
